Registrationless checkout: Add email field validation

niranjan-uma-shankar opened the pull request:
### Changes proposed in this Pull Request

**This is one among a set of stacked PRs that implements userless checkout A/B test. For more reading on the userless
checkout project, read pau2Xa-1hW-p2.**

This PR implements the following:
* Connects the email field to the validation endpoint that verifies if the entered email address is valid.
* Displays an error message below the email field in case of an error. 

### NOTE:
- Depends on D45659-code.
- This PR is a partial implementation only. You will not be able to process the transaction yet, subsequent PRs handle this.

### Testing instructions

#### NOTE: Depends on D45659-code, so please apply this first.

**Scenario 1 - Test invalid email formats**

* Navigate to http://calypso.localhost:3000/, set your locale/geo to EN-US, and assign yourself to _variantUserless_ of `userlessCheckout` test and _control_ of `newSiteGutenbergOnboarding` test.
* Go to /start, select a paid plan and proceed to checkout. To test email validation, you will need to type something in the email field and click "Continue".
* In the email field, try invalid email formats and verify that you are shown an error - e.g `rand@something`, `rand#gmail.com` etc.
* Leave the email field blank and verify you get an error.
* Repeat above steps with a domain in cart - email validation should work in both Billing Info section and Contact Info section(check screenshots in https://github.com/Automattic/wp-calypso/pull/44211).

**Scenario 2 - Test email taken message**
* Navigate to http://calypso.localhost:3000/, set your locale/geo to EN-US, and assign yourself to _variantUserless_ of `userlessCheckout` test and _control_ of `newSiteGutenbergOnboarding` test.
* Go to /start, select a paid plan and proceed to checkout.
* Type an email ID that is tied to an existing wpcom account. You should get a message. You will see an error message - "That email address is already in use. If you have an existing account, please log in.".
* Verify both in the Billing Info and Contact Info sections (i.e without and with domain in cart).
* _Note - Clicking the login link will not work as expected yet. This will be fixed in https://github.com/Automattic/wp-calypso/pull/44238_

sirbrillig commented on client/my-sites/checkout/composite-checkout/contact-validation.js:
```diff
@@ -33,6 +35,14 @@ const wpcomValidateDomainContactInformation = ( ...args ) =>
 	} );
+async function wpcomValidateSignupEmail( ...args ) {
+	try {
+		const response = await wpcom.validateNewUser( ...args, null );
```
It's not very important, but I _think_ that you could change this function to:

```js
async function wpcomValidateSignupEmail( ...args ) {
  return wpcom.validateNewUser( ...args, null );
}
```

...since if `wpcom.validateNewUser` rejects or throws, that will cause this function to reject, and because `throw` in an async function is the same as a rejection.

niranjan-uma-shankar replied:
Good point, I've made this change.

sirbrillig commented on client/my-sites/checkout/composite-checkout/wpcom/components/wp-checkout.js:
```diff
@@ -127,8 +130,46 @@ export default function WPCheckout( {
+		const isNative = config.isEnabled( 'login/native-login-links' );
+		const loginUrl = login( { redirectTo, emailAddress, isNative } );
+		const loginRedirectMessage = translate(
+			'Choose a different email address. This one is not available. If this is you {{a}}log in now{{/a}}.',
```
Minor, but I think the wording of this message could be improved. Maybe, "That email address is already in use. If you have an existing account, {{a}}please log in{{/a}}."

niranjan-uma-shankar replied:
Updated copy.

sirbrillig commented on client/my-sites/checkout/composite-checkout/wpcom/components/wp-checkout.js:
```diff
@@ -127,8 +130,46 @@ export default function WPCheckout( {
 	const validateContactDetailsAndDisplayErrors = async () => {
 		debug( 'validating contact details with side effects' );
+		if ( isLoggedOutCart ) {
+			const email = contactInfo.email?.value;
```
If the optional chaining falls through, then `email` will be `undefined` here, and that `undefined` will be passed through the rest of the functions above. That might actually be totally fine, but if it's not, we could set a fallback value here like:

```js
const email = contactInfo.email?.value ?? '';
```

I don't know if it's an issue so I'll leave it to you to decide.

niranjan-uma-shankar replied:
Ah yes, it doesn't lead to an error but good to have it this way. I've made this change.

sirbrillig commented on client/my-sites/checkout/composite-checkout/wpcom/components/wp-checkout.js:
```diff
@@ -127,8 +130,46 @@ export default function WPCheckout( {
+
 	const validateContactDetailsAndDisplayErrors = async () => {
 		debug( 'validating contact details with side effects' );
+		if ( isLoggedOutCart ) {
```
I know this code isn't very DRY right now, but do we also need to use this block in `validateContactDetails` below? The difference between the functions is that `validateContactDetails` performs the validation and gets a response, but `validateContactDetailsAndDisplayErrors` also displays any errors. (`validateContactDetails` is used when we are checking existing saved contact details so that the contact details step might be skipped.)

niranjan-uma-shankar replied:
I've added `getSignupEmailValidationResult` to the `validateContactDetails ` block

sirbrillig commented on client/my-sites/checkout/composite-checkout/wpcom/types/backend/signup-validation-endpoint.ts:
```diff
@@ -0,0 +1,16 @@
+ * @see WPCOM_JSON_API_Signups_Validation_User_Endpoint
+ */
+
+export type SignupValidationResponse = {
```
👌

nbloomf commented:
Here are the results of my testing:

Scenario 1:
- Using a fresh incognito window, I navigated to `calypso.localhost:3000` and manually assigned myself to `variantUserless`.
- I navigated to `/start`, selected a free subdomain and personal plan, and proceeded to checkout.
- I verified that validation prevented me from proceeding to the next step with the following values for my email address: the empty string, `a`, `0`, `@`, `@.`, `foo@`, `@example`, `@example.com`, `foo@example`, `foo@example..com`, `.@example.com`, `@@example.com`, `baz@example.com.` (trailing period), `foo..bar@example.com` (consecutive periods). The following addresses do validate (as expected): `+@example.com`, `foo.bar@example.com`, `foo+bar@example.com`.
- I repeated the above steps with a paid domain and personal plan in the cart.
- I repeated the above steps with a domain mapping and personal plan in the cart.
- I repeated the above steps with a domain transfer and personal plan in the cart.

Scenario 2:
- Using a fresh incognito window, I navigated to `calypso.localhost:3000` and manually assigned myself to `variantUserless`.
- I navigated to `/start`, selected a paid domain and personal plan, and proceeded to checkout.
- In the contact details step of checkout, I tried to validate an email address associated to an existing wpcom account.
    - The contact details did not validate, and the email field shows the error "That email address is already in use. If you have an existing account please log in."
- I repeated the above steps with a paid domain and personal plan in my cart, with the same results.

One more comment about copy-

> Enter a working email address, so you can receive our messages.

Personally, when a website asks for my email address just so it can send me messages, I'm not inclined to cooperate (even in checkouts). I wonder if this could say something a little more compelling like "so we can send you a receipt" or "so you can manage your purchase later"?

niranjan-uma-shankar commented:
> Personally, when a website asks for my email address just so it can send me messages, I'm not inclined to cooperate (even in checkouts). I wonder if this could say something a little more compelling like "so we can send you a receipt" or "so you can manage your purchase later"?

Thanks for suggesting that. Deferring copy related inputs to @victorespigares.

victorespigares commented:
> > Personally, when a website asks for my email address just so it can send me messages, I'm not inclined to cooperate (even in checkouts). I wonder if this could say something a little more compelling like "so we can send you a receipt" or "so you can manage your purchase later"?
> 
> Thanks for suggesting that. Deferring copy related inputs to @victorespigares.

Thanks @niranjan-uma-shankar for the extra context and screenshot here!
![image](https://user-images.githubusercontent.com/538790/88397557-64ffef00-cdc4-11ea-8ed1-b7677f43b6cc.png)

Since this is a validation error message, I agree it could be misleading: they'll be using that email address to access their account, not just to receive messages. 

I suggest we change it to:
`Use a working email address, so you can access your account.`

niranjan-uma-shankar commented:
> Use a working email address, so you can access your account.

I've updated the copy to reflect this. Also, I noticed that if the email field is left empty, then the error message(from the backend) shown is `Enter a working email address, so you can access your account.`("use" is replaced with "enter"), so I've modified the copy for that too.
